### PR TITLE
[FLINK-13382] [Table SQL / Planner] Port DecimalITCase to unit tests

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

Reduce Blink planner testing time by replacing decimal ITCase  with unit tests.

## Brief change log

- Delete 55 tests from org.apache.flink.table.runtime.batch.sql.DecimalITCase and  45 tests org.apache.flink.table.runtime.batch.table.DecimalITCase, all of these tests can replace with table API tests.
- Add corresponding API tests in org.apache.flink.table.expressions.DecimalTypeTest.
- Add testSqlCalResNullable()、testTableCalResNullable()function in ExpressionTestBase for testing the result of expression(eg f0 +f1) is nullable or not rather than using 'CASE WHEN (f0 +f1) is null THEN 'null' ELSE 'not null' END' in function testSqlNullable() and testTableNullable() . Function testSqlNullable() and testTableNullable() were turned out to be wrong in some cases in FLINK-14030

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no )
  - The S3 file system connector: ( no )

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala [hunk 1]
```diff
@@ -295,32 +310,1042 @@ class DecimalTypeTest extends ExpressionTestBase {
...
+      "f46 % f47",
+      "mod(f46, f47)",
+      "3.12")
+  }
+
+  @Test  // functions that treat Decimal as exact value
+  def testExactionFunctions(): Unit = {
+
+    testAllApis(
+      ExpressionBuilder.ifThenElse('f48 > 'f49, 'f48, 'f49),
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]:
```suggestion
      ifThenElse('f48 > 'f49, 'f48, 'f49),
```

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala [hunk 2]
```diff
@@ -295,32 +310,1042 @@ class DecimalTypeTest extends ExpressionTestBase {
...
+      "cast(f61 as DOUBLE)",
+      "1.99")
+
+    testSqlApi(
+      "cast(f62 as INT)",
+      "1")
+  }
+
+  @Test
+  def testQquality(): Unit = {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: `testEquality`?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala [hunk 3]
```diff
@@ -24,12 +24,27 @@ import org.apache.flink.table.api.{DataTypes, Types}
 import org.apache.flink.table.expressions.utils.ApiExpressionUtils.valueLiteral
 import org.apache.flink.table.planner.expressions.utils.ExpressionTestBase
 import org.apache.flink.table.runtime.typeutils.BigDecimalTypeInfo
+import org.apache.flink.table.runtime.types.TypeInfoLogicalTypeConverter.fromLogicalTypeToTypeInfo
+import org.apache.flink.table.types.logical.DecimalType
 import org.apache.flink.types.Row
-
 import org.junit.{Ignore, Test}
+import java.math.{BigDecimal => JBigDecimal}
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: useless import

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala [hunk 4]
```diff
@@ -295,32 +310,1042 @@ class DecimalTypeTest extends ExpressionTestBase {
...
       "true")
 
-    // TODO add all tests if FLINK-4070 is fixed
     testSqlApi(
       "12 < f1",
       "true")
   }
 
+  @Test
+  def testTableDataSource(): Unit = {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: `testFieldAcess`? There is no DataSource here.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala [hunk 5]
```diff
@@ -295,32 +310,1053 @@ class DecimalTypeTest extends ExpressionTestBase {
       "f4 < f0",
       "true")
 
-    // TODO add all tests if FLINK-4070 is fixed
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: I'll add more tests for this case cause FLINK-4070 has been fixed

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala (hunk 3 above)
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: I will remove it

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala (hunk 2 above)
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: thanks，it's a mistake

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala [hunk 6]
```diff
@@ -295,32 +310,1053 @@ class DecimalTypeTest extends ExpressionTestBase {
...
-      /* 6 */ BigDecimalTypeInfo.of(10, 3))
+      /* 0 */ BigDecimalTypeInfo.of(30, 18),
+      /* 1 */ BigDecimalTypeInfo.of(30, 0),
+      /* 2 */ Types.INT(),
+      /* 3 */ Types.DOUBLE(),
+      /* 4 */ BigDecimalTypeInfo.of(10, 0),
+      /* 5 */ BigDecimalTypeInfo.of(10, 3),
+
+      //convert ITCase to unit Test
+      /* 6 */ fromLogicalTypeToTypeInfo(DECIMAL(10, 0)),
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: That's better style.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala (hunk 4 above)
Comment (leonardBang) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/DecimalTypeTest.scala]: These cases using to test selected column from table source in ITcase, I think ’testFieldAcess' make sense too。

**Comment by leonardBang** [comment]
@wuchong
 I update the PR according to our comments,  I'd appreciate if you can have a look.

**Pull request closed without merging**